include/aws/s3/s3_client.h

```diff
@@ -834,6 +835,74 @@ struct aws_s3_meta_request *aws_s3_client_make_meta_request(
     struct aws_s3_client *client,
     const struct aws_s3_meta_request_options *options);
 
+/**
+ * The result of an `aws_s3_meta_request_poll_write()` call.
+ * Think of this like Rust's `Poll<Result<size_t, int>>`, or C++'s `optional<expected<size_t, int>>`.
+ */
+struct aws_s3_meta_request_poll_write_result {
+    bool is_pending;
+    int error_code;
+    size_t bytes_processed;
+};
+
+/**
+ * Attempt to write data.
+ *
+ * You must set `aws_s3_meta_request_options.send_using_async_writes` to use this function.
+ *
+ * This is a non-blocking poll-style async function, similar to Rust's:
+ * https://docs.rs/futures/latest/futures/io/trait.AsyncWrite.html#tymethod.poll_write
+ * If you prefer completion-style async functions, and your data can outlive
+ * the callstack, use aws_s3_meta_request_write() instead.
+ *
+ * Check the returned `result` struct to see what happened:
+ * 1)   If `result.is_pending == true` then no work was done.
+ *      The waker callback will be invoked when you can call poll_write() again.
+ *      Do not call poll_write() again before the waker is invoked.
+ *
+ * 2)   Else if `result.error_code != 0` then poll_write() did not succeed
```

Comment:
are there any recoverable errors, or error means everything is on fire and the only way forward is to kill things?

Reply:
"result.is_pending" is the only "recoverable" error, which is why I broke it out into its own variable, instead of it making it a special error code.

Full text is:
```
> Else if `result.error_code != 0` then poll_write() did not succeed
> and you should not call it again. The meta request is guaranteed to finish soon
> (you don't need to worry about canceling the meta request yourself after a failed write).
> A common error code is AWS_ERROR_S3_REQUEST_HAS_COMPLETED, indicating
> the meta request completed for reasons unrelated to the poll_write() call
> (e.g. CreateMultipartUpload received a 403 Forbidden response).
> AWS_ERROR_INVALID_STATE usually indicates that you're calling poll_write()
> incorrectly (e.g. not waiting for waker callback from previous poll_write() call).
 ```

My personal philosophy on APIs where users need to call a sequence of functions correctly to complete the operation is: If anything goes wrong with a function in that sequence, the operation should cancel itself. We shouldn't require the user to inspect the result of every code, to determine whether or not they need to cancel the operation. It's a streaming operation, so random failure was always a possibility.

So users can write simple code like:
```
if meta_request.write() failed:
    # something went wrong, bail out of write loop, meta-request will finish soon
    return
```

Instead of:
```
if meta_request.write() failed:
    # something went wrong, be sure to cancel the meta-request
    # in case this is our fault, then bail out of write loop, meta-request will finish soon
    if error_code != AWS_ERROR_S3_REQUEST_HAS_COMPLETED:
        meta_request.cancel()

    return
```